hs_core/views/__init__.py

```diff
@@ -246,12 +248,39 @@ def delete_file(request, shortkey, f, *args, **kwargs):
 def delete_resource(request, shortkey, *args, **kwargs):
     res, _, _ = authorize(request, shortkey, needed_permission=ACTION_TO_AUTHORIZE.DELETE_RESOURCE)
 
+    # downgrade all collections to private if the res being deleted here is the only member resource in these collections
+    # reverse lookup: find all collections that hold this res being deleted
+    associated_collection_metadata_obj_list = list(res.collection_set.all())
+    for collection_metadata_obj in associated_collection_metadata_obj_list:
+        # if this collection has only one member resource (that is the one being deleted now)
+        if collection_metadata_obj.resources.all().count() == 1:
+            # reverse lookup: metadata obj --> resource obj
+            res_collection = CollectionResource.objects.get(object_id=collection_metadata_obj.object_id)
+            if res_collection.raccess.public or res_collection.raccess.discoverable:
+                # downgrade these collections to private
+                res_collection.raccess.public = False
+                res_collection.raccess.discoverable = False
+                res_collection.raccess.save()
+                # set isPublic metadata AVU accordingly
+                istorage = IrodsStorage()
+                istorage.setAVU(res_collection.short_id, "isPublic", str(res_collection.raccess.public))
+
     try:
         hydroshare.delete_resource(shortkey)
     except ValidationError as ex:
         request.session['validation_error'] = ex.message
         return HttpResponseRedirect(request.META['HTTP_REFERER'])
 
+    # the ManyToMany Relationships between this resource and all associated collections have been removed
+    # that means the metadata of these collection also changed, so we need to update their bags
+    for collection_metadata_obj in associated_collection_metadata_obj_list:
+        # tricky issue: the user who deleted this resource may not have permission to access this collection
```

Comment:
comment on (2): we are using ManyToMany relationship to associate a collection with its contained resources (rather than storing their res_ids in collection), which is recommend by the developer team. Say, collectionA holds resB, that means collectionA object is referencing resB obj. if resB obj gets removed, the tie between colletionA and resB will be gone automatically...so there is no broken link in collection A at all....